validator-network/src/network_impl.rs

```diff
@@ -76,50 +87,116 @@ where
+            }
+        };
+        self.validator_peer_id_cache
+            .write()
+            .insert(public_key.compressed().clone(), cache_value);
+    }
+
     /// Look up the peer ID for a validator ID.
-    async fn get_validator_peer_id(
+    fn get_validator_peer_id(
         &self,
         validator_id: u16,
     ) -> Result<N::PeerId, NetworkError<N::Error>> {
-        let (peer_id, public_key) = {
-            let public_key = self
-                .validator_keys
-                .read()
-                .get(usize::from(validator_id))
-                .ok_or(NetworkError::UnknownValidator(validator_id))?
-                .clone();
+        let public_key = self
+            .validator_keys
+            .read()
+            .get(usize::from(validator_id))
+            .ok_or(NetworkError::UnknownValidator(validator_id))?
+            .clone();
 
-            let peer_id = self
+        let update_peer_cache = {
+            if let Some(cache_state) = self
                 .validator_peer_id_cache
                 .read()
                 .get(&public_key.compressed().clone())
-                .cloned();
-            (peer_id, public_key)
+            {
+                match cache_state {
+                    CacheState::Resolved(peer_id) => return Ok(*peer_id),
+                    CacheState::Error(_) => true,
+                    CacheState::InProgress => {
+                        log::debug!(validator_id, "Record resolution is in progress");
+                        false
+                    }
+                }
+            } else {
+                log::debug!(
+                    ?public_key,
+                    validator_id,
+                    "Could not find peer ID for validator in DHT. Performing a DHT query",
+                );
+                true
+            }
         };
-
-        if let Some(peer_id) = peer_id {
-            Ok(peer_id)
-        } else if let Some(peer_id) = Self::resolve_peer_id(&self.network, &public_key).await? {
-            self.validator_peer_id_cache
-                .write()
-                .insert(public_key.compressed().clone(), peer_id);
-            Ok(peer_id)
-        } else {
-            log::error!(
-                "Could not find peer ID for validator in DHT: public_key = {:?}",
-                public_key
-            );
-            Err(NetworkError::UnknownValidator(validator_id))
+        if update_peer_cache {
+            // Cache is empty for this validator ID, query the entry
+            {
+                // Re-check the validator Peer ID cache with the write lock taken and update it if necessary
+                let mut validator_peer_id_cache = self.validator_peer_id_cache.write();
+                if let Some(cache_state) =
+                    validator_peer_id_cache.get(&public_key.compressed().clone())
+                {
+                    match cache_state {
+                        CacheState::Resolved(_) => {}
```

Comment:
```suggestion
                        CacheState::Resolved(peer_id) => return Ok(*peer_id),
```
In this case, another thread has already resolved it. We can return the result immediately.